File: check_divs.js
const fs = require('fs');
const content = fs.readFileSync('components/WorkspaceStudio.tsx', 'utf8');

const returnRegex = /return \([\s\S]+?\);\n\};/m;
const match = content.match(returnRegex);

if (match) {
    const block = match[0];
    const openDivs = (block.match(/<div\b/g) || []).length;
    const closeDivs = (block.match(/<\/div\s*>/g) || []).length;
    console.log(`Open: ${openDivs}, Close: ${closeDivs}`);
    
    // Let's print out every div log to see where they fail to match
    let depth = 0;
    const lines = block.split('\n');
    lines.forEach((line, i) => {
        const opens = (line.match(/<div\b/g) || []).length;
        const closes = (line.match(/<\/div\s*>/g) || []).length;
        if (opens > 0 || closes > 0) {
            depth += (opens - closes);
            console.log(`${i+1}: OP:${opens} CL:${closes} -> Depth:${depth} | ${line.trim()}`);
        }
    });

} else {
    console.log("No return block found!");
}

File: check_divs.py
import re

with open("components/WorkspaceStudio.tsx", "r", encoding="utf-8") as f:
    content = f.read()

# find return block
match = re.search(r'return \([\s\S]+?\);\n\};', content)
if match:
    block = match.group(0)
    lines = block.split("\n")
    depth = 0
    for i, line in enumerate(lines):
        opens = len(re.findall(r'<div\b', line))
        closes = len(re.findall(r'</div\s*>', line))
        if opens or closes:
            depth += (opens - closes)
            print(f"Line {i+354}: +{opens} -{closes} Depth: {depth} | {line.strip()}")
        
        o_frags = len(re.findall(r'<>', line))
        c_frags = len(re.findall(r'</>', line))
        if o_frags or c_frags:
            print(f"Line {i+354}: FRAG +{o_frags} -{c_frags} | {line.strip()}")
else:
    print("NO MATCH")
